### PR TITLE
Refactoring of Deadline to ReminderDate

### PR DESCRIPTION
**Pull request by kerbysoh**
- Refactor class name of Deadline to ReminderDate to more clearly highlight the purpose of the field within the role.

- ReminderDate is now an optional field within the addRoleCommand. When users opt not to key in a reminder date, the ReminderDate field within the Role would be created with its value field set to null.

- A ReminderDate with a null value field is a valid ReminderDate.

- Updated storage to parse null-valued ReminderDate as empty strings

- Update UI to display "N/A" when any role field is empty.

- Update UG, DG and relevant UML diagrams with the newly refactored names

### CODE REVIEW

**Comment by kerbysoh** [comment]
Possible grammatical and wording errors where deadline is more appropriate than reminder date.

File: docs/UserGuide.md [hunk 1]
```diff
@@ -245,8 +247,6 @@ Colour coding of roles allows users to quickly obtain information about their ap
 * "offered": green
 * "complete": black
 
-Format: `exit`
-
```
Comment (petermonky) [inline docs/UserGuide.md:249]: Nice fix here 👍

File: src/main/java/seedu/tinner/logic/parser/CliSyntax.java [hunk 2]
```diff
@@ -13,7 +13,7 @@
     public static final Prefix PREFIX_EMAIL = new Prefix("e/");
     public static final Prefix PREFIX_ADDRESS = new Prefix("a/");
     public static final Prefix PREFIX_STATUS = new Prefix("s/");
-    public static final Prefix PREFIX_DEADLINE = new Prefix("b/");
+    public static final Prefix PREFIX_REMINDER_DATE = new Prefix("b/");
```
Comment (petermonky) [inline src/main/java/seedu/tinner/logic/parser/CliSyntax.java]: Perhaps we can change the prefix to `r/`? And somehow unify both company name and role name prefixes as `n/`.

**Review by petermonky — APPROVED** [review]
Clean refactoring 👍 LGTM

File: src/main/java/seedu/tinner/logic/parser/CliSyntax.java (hunk 2 above)
Comment (justinyjt) [reply src/main/java/seedu/tinner/logic/parser/CliSyntax.java]: >Perhaps we can change the prefix to r/

I agree with Peter on this, makes more intuitive sense for the user if reminder date had the r/ prefix.

>unify both company name and role name prefixes as n/

I believe that these prefixes are for the find command where you need to specify both company and role name in the same command, hence the distinction between the two.

File: src/main/java/seedu/tinner/model/role/ReminderDate.java [hunk 3]
```diff
@@ -53,35 +60,46 @@ public static boolean isValidDeadline(String test) {
...
+     * Returns true if the reminder date is within a week from the current {@code LocalDateTime}
+     * or if the date is null.
      *
      * @return true if is within a week from the current date.
      */
     public boolean isWithinReminderWindow() {
+        if (this.value == null) {
+            return false;
+        }
         LocalDateTime oneWeekAway = LocalDateTime.now().plusDays(reminderWindow);
```
Comment (justinyjt) [inline src/main/java/seedu/tinner/model/role/ReminderDate.java]: Spotted a mini bug of my own here. I forgot to change the LocalDateTime variable name to oneReminderWindowAway. Could you help to just make a quick change to this for me so I don't need to make a whole PR just for that one variable 😅 . Thanks!

File: src/main/java/seedu/tinner/storage/JsonAdaptedRole.java [hunk 4]
```diff
@@ -80,16 +84,14 @@ public Role toModelType() throws IllegalValueException {
         }
         final Status roleStatus = new Status(status);
 
-        if (deadline == null) {
+        if (reminderDate == null) {
```
Comment (justinyjt) [inline src/main/java/seedu/tinner/storage/JsonAdaptedRole.java:87]: Based on our previous implementation, we removed null checks for optional fields as it is possible for these fields to be null. Should this apply to reminder date as well?

File: src/test/java/seedu/tinner/storage/JsonAdaptedRoleTest.java [hunk 5]
```diff
@@ -56,47 +57,47 @@ public void toModelType_nullRoleName_throwsIllegalValueException() {
...
-                new JsonAdaptedRole(VALID_NAME, VALID_STATUS, INVALID_DEADLINE, VALID_DESCRIPTION, VALID_STIPEND);
-        String expectedMessage = Deadline.MESSAGE_CONSTRAINTS;
+                new JsonAdaptedRole(VALID_NAME, VALID_STATUS, INVALID_REMINDER_DATE, VALID_DESCRIPTION, VALID_STIPEND);
+        String expectedMessage = ReminderDate.MESSAGE_CONSTRAINTS;
         assertThrows(IllegalValueException.class, expectedMessage, role::toModelType);
     }
 
     @Test
-    public void toModelType_nullDeadline_throwsIllegalValueException() {
+    public void toModelType_nullReminderDate_throwsIllegalValueException() {
```
Comment (justinyjt) [inline src/test/java/seedu/tinner/storage/JsonAdaptedRoleTest.java:82]: Refer to previous comment on JsonAdaptedRole: this test case may not be required.

**Review by justinyjt — COMMENTED** [review]
Overall LGTM. Looks like you managed to capture most of the changes despite this being quite a large refactoring. Nice work :)

File: src/main/java/seedu/tinner/model/role/ReminderDate.java (hunk 3 above)
Comment (kerbysoh) [reply src/main/java/seedu/tinner/model/role/ReminderDate.java]: Sure no worries

File: src/main/java/seedu/tinner/storage/JsonAdaptedRole.java (hunk 4 above)
Comment (kerbysoh) [reply src/main/java/seedu/tinner/storage/JsonAdaptedRole.java:87]: Can someone weigh in on this? @petermonky @BrendonLau @jinnhl? I believe that with our current implementation none of those optional fields will be null. They are empty strings but not null values. Correct me if I'm wrong.

File: src/main/java/seedu/tinner/logic/parser/CliSyntax.java (hunk 2 above)
Comment (kerbysoh) [reply src/main/java/seedu/tinner/logic/parser/CliSyntax.java]: I agree. Will make relevant changes.

Comment (petermonky) [reply src/main/java/seedu/tinner/logic/parser/CliSyntax.java]: > I believe that these prefixes are for the find command where you need to specify both company and role name in the same command, hence the distinction between the two.


Ahhh yep forgot about the find command, using `n/` for both could definitely pose an issue then. Perhaps could use some other prefixes then.

File: src/main/java/seedu/tinner/storage/JsonAdaptedRole.java (hunk 4 above)
Comment (petermonky) [reply src/main/java/seedu/tinner/storage/JsonAdaptedRole.java:87]: > I believe that with our current implementation none of those optional fields will be null.

Does this apply to other classes as well? If so then I think we can add `null`checks the other classes too, just to be more defensive.

Comment (kerbysoh) [reply src/main/java/seedu/tinner/storage/JsonAdaptedRole.java:87]: Yes I have updated null checks for them. Testing has yet to be updated.

File: src/main/java/seedu/tinner/logic/parser/CliSyntax.java (hunk 2 above)
Comment (kerbysoh) [reply src/main/java/seedu/tinner/logic/parser/CliSyntax.java]: Yes I have retained the use of prefixes for the find command using c/ and r/. Only the addRole and editRole command has been edited, where the reminderDate field is indicated with a prefix r/.

**Pull request merged**